api/wasm/cpp/proxy_wasm_api.h

```diff
@@ -476,31 +476,44 @@ inline Optional<WasmDataPtr> getProperty(std::initializer_list<StringView> parts
   return std::make_unique<WasmData>(value_ptr, value_size);
 }
 
-inline bool getStringValue(std::initializer_list<StringView> parts, std::string* out) {
+// Generic property reader for basic types: int64, uint64, double, bool
+// Durations are represented as int64 nanoseconds.
+// Timestamps are represented as int64 Unix nanoseconds.
+// Strings and bytes are represented as std::string.
+template <typename T>
+inline bool getValue(std::initializer_list<StringView> parts, T* out) {
   auto buf = getProperty(parts);
-  if (!buf.has_value()) {
+  if (!buf.has_value() || buf.value()->size() != sizeof(T)) {
     return false;
   }
-  out->assign(buf.value()->data(), buf.value()->size());
+  *out = *reinterpret_cast<const T*>(buf.value()->data());
   return true;
 }
 
-inline bool getStructValue(std::initializer_list<StringView> parts,
-                           google::protobuf::Struct* value_ptr) {
+// Specialization for bytes and string values
+template <>
+inline bool getValue<std::string>(std::initializer_list<StringView> parts, std::string* out) {
   auto buf = getProperty(parts);
   if (!buf.has_value()) {
     return false;
   }
-  return value_ptr->ParseFromArray(buf.value()->data(), buf.value()->size());
+  out->assign(buf.value()->data(), buf.value()->size());
+  return true;
 }
 
-template <typename T> inline bool getValue(std::initializer_list<StringView> parts, T* out) {
+// Specialization for message types (including struct value for lists and maps)
+template <typename T>
+inline bool getMessageValue(std::initializer_list<StringView> parts,
+                            T* value_ptr) {
   auto buf = getProperty(parts);
-  if (!buf.has_value() || buf.value()->size() != sizeof(T)) {
+  if (!buf.has_value()) {
     return false;
   }
-  *out = *reinterpret_cast<const T*>(buf.value()->data());
-  return true;
+  if (buf.value()->size() == 0) {
```

Comment:
Should this return a pointer to empty T instead of a nullptr? What does ParseFromArray do when size is 0?

Reply:
Zero array is intentional if you look into the context. nullptr is a valid value that is reported as a zero-size message. ParseFromArray is probably going to return an empty message which is not right, since it's usually implemented via merge.